Product Collection: Integrate Customer Effort Score Feedback Modal

imanish003 opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #52730

This PR adds Customer Effort Score to Product Collection block.
- It will show "Help us improve" button at bottom of Inspector controls.
- When clicked, it will show a Modal asking for feedback about their experience with the block, including:
  - A rating question: "It was easy for me to accomplish what I wanted with the Product Collection block"
  - An optional feedback text field asking how we can improve the block
  - An optional email field for follow-up research participation
- On submission, it will show a success notice thanking them for their feedback

#### Reusable Customer Effort Score Button Component

This PR also adds a reusable CES feedback button component that can be used across different blocks:

1. Created `CesFeedbackButton` component in `editor-components/ces-feedback-button/index.tsx`
2. Component accepts props to customize:
   - Block name
   - Modal title
   - Questions and labels
   - Button text
   - Submit message
   - Optional wrapper component

##### Props

- `blockName` (string, required)
  - Name of the block to collect feedback for
  - Used in questions and tracking

- `title` (string, optional) 
  - Title shown at the top of the feedback modal
  - Default: "Share your experience"

- `firstQuestion` (string, optional)
  - The main rating question shown in the modal
  - Default: "It was easy for me to accomplish what I wanted with the [blockName]."

- `feedbackLabel` (string, optional)
  - Label for the feedback text area
  - Default: "How can we improve the [blockName] block for you? (Optional)"

- `feedbackPlaceholder` (string, optional)
  - Placeholder text shown in the feedback text area
  - Default: "What did you try to build using this block? What did and didn't work?"

- `emailLabel` (string, optional)
  - Label for the email input field
  - Default: "Email address (Optional)"

- `emailHelp` (string, optional)
  - Help text shown below the email field
  - Default: "Share if you would like to discuss your experience or participate in future research."

- `buttonText` (string, optional)
  - Text shown on the feedback button
  - Default: "Help us improve"

- `submitLabel` (string, optional)
  - Message shown after feedback submission
  - Default: "🙏🏻 Thanks for sharing — we're on it!"

- `wrapper` (React.ElementType, optional)
  - Component to wrap the feedback button. The wrapper component is only rendered when WooCommerce Tracks is enabled (`window.wcTracks?.isEnabled`). This allows us to conditionally show/hide both the feedback button and its wrapper based on tracking settings.
  - Example: `PanelBody` from WordPress components

- `wrapperProps` (object, optional)
  - Props to pass to the wrapper component
  - Default: `{}`

**Usage**

```tsx
import { CesFeedbackButton } from '@woocommerce/editor-components';

<CesFeedbackButton blockName="Product Collection" />
```

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

**Preparations:**
1. Create new post
2. Open your browser Dev Tools
3. In Console tab paste this code and submit by enter:
```js
localStorage.setItem( 'debug', 'wc-admin:*' )
```
4. Enable "Verbose" logging level

**Testing:**
1. Add "Product Collection" block to a new post
2. Choose any collection
3. You will notice a button "Help us improve" at bottom of Inspector settings
<img src="https://github.com/user-attachments/assets/9b4e6868-6600-496a-80c9-925876728720" width="200" />


4. Click the button to open the feedback form. Here is how it looks:
<img src="https://github.com/user-attachments/assets/f9a364ff-264d-4670-a216-805fc04d7c01" width="600" />

5. Attempt to submit the feedback form. Note that while the rating question is required, the other fields are optional. If you try to submit without selecting a rating, you should see an error message indicating this is a required field. 

6. Select any rating option and submit. Verify in console that the feedback form is submitted successfully with following data:
	- Event name: `wcadmin_ces_feedback`
	- Event properties:
		- `action`: "product_collection_block_feedback"
		- `feedback_comment`: Entered feedback text (if any)
		- `score`: Selected rating value (1-5)
		- `store_age`: How long the store has been installed, measured in weeks

![image](https://github.com/user-attachments/assets/2fdbdb1a-2d84-44e0-b648-16d95755a9ee)


7. The following success notice will be shown in the top of the editor:
![Image](https://github.com/user-attachments/assets/a670fa43-0102-4fb4-b720-a4d6976a668c)


8. Try filling out all fields in the form and verify that all entered data is included in the event properties when submitted:
   - Select a rating (1-5)
   - Enter feedback in the comments field
   - Submit the form and check the console to confirm all fields are captured correctly in the `wcadmin_ces_feedback` event

9. Test email validation:
   - Enter an invalid email address and try to submit
   - Verify that an error message appears: "Please enter a valid email address."
   - Enter a valid email address and verify the form submits successfully

10. Verify that the feedback button only appears when WC Tracks is enabled:
    - Disable WC Tracks by running in console: `window.wcTracks.isEnabled = false`
    - Verify the feedback button disappears
        - Note: After changing `window.wcTracks.isEnabled`, avoid refreshing the page as it will reset the value. To verify the change took effect, click elsewhere in the editor and select the block again.
    - Re-enable WC Tracks: `window.wcTracks.isEnabled = true`
    - Verify the feedback button reappears

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [x] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Product Collection: Add Customer Effort Score feedback form
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

imanish003 committed:
Add feedback button to Product Collection inspector controls

Add a "Help us improve" button to the Product Collection block's inspector controls panel.
The button uses a new feedback icon and will be used to gather user feedback.

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
First working version of Customer Effort Score feedback modal

- Add feedback button component with CES modal integration
- Add styling for error messages in CES modal
- Register and enqueue CES and experimental scripts/styles
- Update package dependencies to include experimental package

Note: Build error exists and will be fixed in a follow-up commit

imanish003 committed:
Fix build issue

imanish003 committed:
Add @woocommerce/customer-effort-score to webpack dependency maps

This commit adds the necessary webpack configuration to support the
@woocommerce/customer-effort-score package in the WooCommerce Blocks
project by adding it to both wcDepMap and wcHandleMap. This allows us
to use the package without needing to build it separately.

imanish003 committed:
Move CES feedback button to Editor Components

This commit moves the Customer Effort Score feedback button from the Product Collection block to a reusable component in the editor-components directory. This will allow us to reuse the feedback button across other blocks in the future.

The new component is more configurable, allowing customization of all text strings while providing sensible defaults. The action name is now automatically generated from the block name.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

imanish003 committed:
Add feedback icon for CES feedback button

imanish003 committed:
Merge branch 'add/52730-product-collection-integrate-customer-effort-score-feedback-modal' of https://github.com/woocommerce/woocommerce into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
Match the design of button using CSS

imanish003 committed:
refactor: Remove unnecessary fragment in inspector controls

imanish003 committed:
Regenerate lock file

imanish003 committed:
Merge branch 'trunk' into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
Remove old prompt and change position of new one

imanish003 committed:
move shouldShowComments to 2nd argument of showCesModal

kmanijak commented on plugins/woocommerce-blocks/assets/css/editor.scss:
```diff
@@ -28,11 +28,11 @@
 .components-base-control {
-	+ .wc-block-base-control-notice {
+	+.wc-block-base-control-notice {
 		margin: -$gap 0 $gap;
 	}
 
-	+ .wc-block-base-control-notice:last-child {
+	+.wc-block-base-control-notice:last-child {
```
Is this change just cosmetic or it has some implications? AFAIK it would compile to the same CSS so I'm curious.

imanish003 replied:
> Is this change just cosmetic or it has some implications? AFAIK it would compile to the same CSS so I'm curious.

It's just cosmetic. When I saved the file, it automatically formatted the CSS. I thought I could keep the auto-formatted code. Is it auto-formatting to another format when you save the file? I think auto-formatting is following the `.markdownlint.json` file. 🤔

imanish003 replied:
I have reverted this change in e9a1151.

kmanijak commented on plugins/woocommerce-blocks/assets/css/editor.scss:
```diff
@@ -51,6 +51,7 @@ svg.wc-block-editor-components-block-icon--sparkles path {
 
 .theme-twentytwenty {
 	.wp-block {
+
```
Unnecessary empty line

```suggestion
```

imanish003 replied:
Again this was auto-formatted. Probably, My auto-formatter is not working as expected? 😅

imanish003 replied:
I have reverted this change in e9a1151.

kmanijak commented on plugins/woocommerce-blocks/assets/css/editor.scss:
```diff
@@ -60,3 +61,9 @@ svg.wc-block-editor-components-block-icon--sparkles path {
+
+// Style the error message in the CES modal.
+.woocommerce-customer-effort-score__errors > p {
+	color: #d94f4f;
```
I see it's a the same colour as `$error-red: #d94f4f;` from https://github.com/woocommerce/woocommerce/blob/6cf227d2cb5cfbf49dc5d1b8b9d5cd30648273b8/packages/js/internal-style-build/abstracts/_colors.scss#L24.

Nitpick, but maybe we can leave a comment why we use such a hardcoded colour. I don't have a strong opinion though 😅

imanish003 replied:
Good catch! It doesn't make sense to use hardcoded color. I have decided to go with `$alert-red` color which we are importing from [@wordpress/base-styles/colors.scss](https://github.com/woocommerce/woocommerce/blob/828fb38f800896a66cd0ff7117b81c56b1623228/packages/js/internal-style-build/abstracts/_variables.scss#L4). Does this sound good?

kmanijak commented on plugins/woocommerce-blocks/assets/js/editor-components/ces-feedback-button/index.tsx:
```diff
@@ -0,0 +1,123 @@
+	blockName,
+	title = __( 'Share your experience', 'woocommerce' ),
+	// translators: %s is the block name.
+	firstQuestion = __(
+		'It was easy for me to accomplish what I wanted with the %s.',
+		'woocommerce'
+	).replace( '%s', blockName ),
+	// translators: %s is the block name.
+	feedbackLabel = __(
+		`How can we improve the %s block for you? (Optional)`,
+		'woocommerce'
+	).replace( '%s', blockName ),
```
I think we should use `sprintf` function from `@wordpress/i18n`

```suggestion
	firstQuestion = sprintf( __(
		'It was easy for me to accomplish what I wanted with the %s.',
		'woocommerce'
	), blockName ),
	// translators: %s is the block name.
	feedbackLabel = sprintf( __
		`How can we improve the %s block for you? (Optional)`,
		'woocommerce'
	), blockName ),
```

kmanijak reviewed (approved):
Thanks for taking care of it. I'm glad to see a unified form we can use across all blocks!

Feedback modal works like a charm! I left some minor comments but pre-approving!

imanish003 committed:
Use color variable $alert-red

imanish003 committed:
Use `sprintf` for localized strings

imanish003 committed:
Merge branch 'trunk' into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
Merge branch 'trunk' into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
Merge branch 'trunk' into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
Merge branch 'trunk' into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
Merge branch 'trunk' into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
Regenerate pnpm-lock.yaml

imanish003 committed:
Revert pnpm-lock.yaml

imanish003 committed:
Regenerate lockfile

imanish003 committed:
Merge branch 'trunk' into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
Merge branch 'trunk' into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
Merge branch 'trunk' into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
Reset pnpm lock file

imanish003 committed:
Merge branch 'add/52730-product-collection-integrate-customer-effort-score-feedback-modal' of https://github.com/woocommerce/woocommerce into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
Merge branch 'trunk' into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
Regenerate lock file

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
Revert pnpm lock file

imanish003 committed:
Regenerate lock file

imanish003 committed:
Merge branch 'trunk' into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
Revert pnpm lock file

imanish003 committed:
Regenerate lock file

imanish003 committed:
Try removing the dependency

imanish003 committed:
Merge branch 'trunk' into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
Add dependencies for Customer Effort Score

imanish003 committed:
Remove wc-experimental dependency from customer effort score

imanish003 commented:
Hi Karol, Today I spent time unblocking this PR and trying to get it merged. I have [removed](https://github.com/woocommerce/woocommerce/pull/52874/commits/31960d611c15124b469011740e128e563ea3ad9b) the `wc-experimental` dependency since it was not required and was causing E2E test failures. Could you please do another quick review before I proceed with merging this PR? I have done some testing, but it would be great if you could review it again.


Sidenote: I added `wc-experimental` dependency because Customer Effort Score wasn't working without it. Maybe I had some build issue locally. But looks like it's not required, therefore I removed it.

kmanijak reviewed (commented):
Code-wise all looks good! Also, I tested the changes and events appear correctly in the console 💪 


But I'm getting this error in console when submitting feedback:

```post.php?post=567&action=edit:2093 A valid prop name must be specified. The property name: "feedbackComment" is not valid.```

<img width="1142" alt="image" src="https://github.com/user-attachments/assets/527d3309-dbec-4bec-be93-4b196f76d05b" />

Steps to reproduce:
1. Create new post
2. Add Product Collection and choose "create your own"
3. Try adding some feedback WITH comment

imanish003 committed:
Update property name for feedback comment

The property name `feedbackComment` was failing validation in WC_Tracks_Event
because it used camelCase. The validation regex `^[a-z_][a-z0-9_]*$` requires
property names to:
- Start with a lowercase letter or underscore
- Only contain lowercase letters, numbers, and underscores
- Cannot contain any other characters (like camelCase)

Changed the property name to `feedback_comment` to match the required format.

imanish003 committed:
Regenerate lock file

imanish003 committed:
Remove changes from pnpm lock file

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into add/52730-product-collection-integrate-customer-effort-score-feedback-modal

imanish003 committed:
Fix merge conflicts with pnpm-lock.yaml

imanish003 commented:
@kmanijak Good catch! After debugging, I found that error is coming from [here](https://github.com/woocommerce/woocommerce/blob/31960d611c15124b469011740e128e563ea3ad9b/plugins/woocommerce/includes/tracks/class-wc-tracks-event.php#L173-L175). There is regex validation for property names which is not allowing camelCase. I have changed the property name from `feedbackComment` to `feedback_comment` to match the required format for WooCommerce Tracks property names. This should resolve the error you were seeing.

FWIW It's using regex `^[a-z_][a-z0-9_]*$` which allows names that:
- Start with a lowercase letter or underscore
- Only contain lowercase letters, numbers, and underscores
- Cannot contain any other characters (like camelCase)

Could you please review the PR again?

kmanijak reviewed (approved):
Works as expected! Thanks for this whole work Manish, it's gonna be great once we propagate this on all blocks!

One improvement I see is some sort of validation to email field. There's `isEmail` function available in `@wordpress/url` which could be used to verify the input in email field is actually an email. And I see there's `validateExtraFields` prop exposed on `CustomerFeedbackModal` but didn't investigate how to use it. If super easy I think it would be a nice addition to this modal. If not, I think it's alright to progress as-is and add it separately.

Having said that, I'm approving the PR in the meantime! Great job!

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/index.tsx:
```diff
@@ -221,7 +221,9 @@ const ProductCollectionInspectorControls = (
 				</ToolsPanel>
 			) : null }
-			<ProductCollectionFeedbackPrompt />
+			<PanelBody>
```
While working on another, unrelated issue regarding tracks I realised we should not be displaying feedback option if user doesn't allow for tracking (if tracking is disabled, feedback won't be sent anyway).

This is the way to check if tracking is enabled:

```
window.wcTracks?.isEnabled
```

and here's example:

https://github.com/woocommerce/woocommerce/blob/3f6b19126e26235ba6df5688632e96f8be6a08f5/packages/js/product-editor/src/hooks/use-feedback-bar/use-feedback-bar.ts#L30

Please add it before merging.

kmanijak reviewed (changes requested):
I'm revoking an approval as I accidentally found out one additional change required. I'm happy to approve once it's done 🙌

imanish003 committed:
Add email validation and check wcTracks?.isEnabled before rendering feedback button

- Added email validation using isEmail utility to ensure valid email addresses
- Added error display for invalid email addresses in feedback form
- Added wcTracks.isEnabled check to conditionally render feedback button
	- Added wrapper prop to CesFeedbackButton for flexible rendering with custom components

imanish003 commented:
@kmanijak Thank you for your thorough review! I've implemented both suggested changes:

1. Added email validation using `isEmail` from `@wordpress/url`
2. Added tracking permission check using `window.wcTracks?.isEnabled` before displaying the feedback button 
   - This required me to add a `wrapper` prop because if I keep the wrapper in the Product Collection block, then it shows empty, as shown in the screenshot below.

![image](https://github.com/user-attachments/assets/5bf668c2-0c14-4fbc-b423-b1740a4ca005)


The changes are ready for another review. Would you mind taking another look? Let me know if you'd like me to explain any part of the implementation in more detail.

imanish003 committed:
Renamed the button variable to feedbackButtonWithModal for better readability

kmanijak reviewed (approved):
Thank you for your patience in addressing all the findings especially that they were raised "in batches" rather than at single review. Amazing job!

🚢 !
